File: app/src/test/java/com/github/fo2rist/mclaren/repository/StoryStreamRepositoryImplTest.java
package com.github.fo2rist.mclaren.repository;

import com.github.fo2rist.mclaren.repository.FeedRepositoryEventBus.LoadingEvent;
import com.github.fo2rist.mclaren.web.FeedWebService.FeedRequestCallback;
import com.github.fo2rist.mclaren.web.StoryStreamWebService;
import java.net.URL;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import static com.github.fo2rist.mclaren.testdata.StoryStreamResponse.REAL_FEED_RESPONSE;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

@RunWith(RobolectricTestRunner.class)
@Config(sdk = 21)
public class StoryStreamRepositoryImplTest {
    //TODO remove tests duplications with MCL Repo test. 2018.11.25

    private FeedRepository repository;
    private StoryStreamWebService mockWebService;
    private FeedRepositoryEventBus mockEventBus;

    @Before
    public void setUp() {
        mockWebService = mock(StoryStreamWebService.class);
        mockEventBus = mock(FeedRepositoryEventBus.class);

        repository = new StoryStreamRepositoryImpl(mockWebService, mockEventBus);
    }

    @Test
    public void test_loadLatest_startLoading_and_firesLoadStartEvent() {
        repository.loadLatestPage();

        verify(mockEventBus).publish(any(LoadingEvent.LoadingStarted.class));
        verify(mockWebService).requestLatestFeed(any(FeedRequestCallback.class));
    }

    @Test
    public void test_loadNextHistory_startLoading_and_firesLoadStartEvent() {
        repository.loadNextPage();

        verify(mockEventBus).publish(any(LoadingEvent.LoadingStarted.class));
        verify(mockWebService).requestFeedPage(anyInt(), any(FeedRequestCallback.class));
    }

    @Test
    public void test_onSuccess_firesLoadFinishSuccessfullyEvents() {
        doAnswer(new Answer() {
            @Override
            public Object answer(InvocationOnMock invocationOnMock) throws Throwable {
                FeedRequestCallback callback = invocationOnMock.getArgument(0);
                callback.onSuccess(new URL("http://empty.url"), 0, 200, REAL_FEED_RESPONSE);
                return null;
            }
        }).when(mockWebService).requestLatestFeed(any(FeedRequestCallback.class));

        repository.loadLatestPage();

        verify(mockEventBus).publish(any(LoadingEvent.LoadingFinished.class));
        verify(mockEventBus).publish(any(LoadingEvent.FeedUpdateReady.class));
    }

    @Test
    public void test_onFailure_firesLoadFinishWithErrorEvents() {
        doAnswer(new Answer() {
            @Override
            public Object answer(InvocationOnMock invocationOnMock) throws Throwable {
                FeedRequestCallback callback = invocationOnMock.getArgument(0);
                callback.onFailure(new URL("http://empty.url"), 0, 400, null);
                return null;
            }
        }).when(mockWebService).requestLatestFeed(any(FeedRequestCallback.class));

        repository.loadLatestPage();

        verify(mockEventBus).publish(any(LoadingEvent.LoadingFinished.class));
        verify(mockEventBus).publish(any(LoadingEvent.LoadingError.class));
    }
}

File: app/src/test/java/com/github/fo2rist/mclaren/repository/StoryStreamRepositoryImplTest.kt
package com.github.fo2rist.mclaren.repository

import com.github.fo2rist.mclaren.repository.FeedRepositoryEventBus.LoadingEvent
import com.github.fo2rist.mclaren.testdata.StoryStreamResponse.REAL_FEED_RESPONSE
import com.github.fo2rist.mclaren.web.StoryStreamWebService
import com.github.fo2rist.mclaren.web.utils.BadResponse
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.doReturn
import com.nhaarman.mockitokotlin2.doThrow
import com.nhaarman.mockitokotlin2.stubbing
import com.nhaarman.mockitokotlin2.verifyBlocking
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.anyInt
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import org.robolectric.RobolectricTestRunner
import java.net.URL

//TODO remove tests duplications with MCL Repo test. 2018.11.25
@RunWith(RobolectricTestRunner::class)
class StoryStreamRepositoryImplTest {

    private lateinit var repository: FeedRepository
    private lateinit var mockWebService: StoryStreamWebService
    private lateinit var mockEventBus: FeedRepositoryEventBus

    @Before
    fun setUp() {
        mockWebService = mock(StoryStreamWebService::class.java)
        mockEventBus = mock(FeedRepositoryEventBus::class.java)

        repository = StoryStreamRepositoryImpl(mockWebService, mockEventBus)
    }

    @Test
    fun test_loadNextHistory_startLoading_and_firesLoadStartEvent() {
        repository.loadNextPage()

        verify(mockEventBus).publish(any<LoadingEvent.LoadingStarted>())
        verifyBlocking(mockWebService) { requestFeedPage(anyInt()) }
    }

    @Test
    fun test_onSuccess_firesLoadFinishSuccessfullyEvents() {
        stubbing(mockWebService) {
            onBlocking { requestLatestFeed() }.doReturn(REAL_FEED_RESPONSE)
        }

        repository.loadLatestPage()

        verify(mockEventBus).publish(any<LoadingEvent.LoadingStarted>())
        verify(mockEventBus).publish(any<LoadingEvent.FeedUpdateReady>())
        verify(mockEventBus).publish(any<LoadingEvent.LoadingFinished>())
    }

    @Test
    fun test_onFailure_firesLoadFinishWithErrorEvents() {
        stubbing(mockWebService) {
            onBlocking { requestLatestFeed() }.doThrow(BadResponse(URL("http://empty.url"), 400))
        }

        repository.loadLatestPage()

        verify(mockEventBus).publish(any<LoadingEvent.LoadingStarted>())
        verify(mockEventBus).publish(any<LoadingEvent.LoadingError>())
        verify(mockEventBus).publish(any<LoadingEvent.LoadingFinished>())
    }
}
